### DISCUSSION TITLE
ArrayField onFieldValueChange 无法正常触发

### DISCUSSION THREAD

**Discussion by prgrmrwy** [alibaba/formily#1922] [Q&A 2.x] [answered]
如 codesandebox: https://codesandbox.io/s/great-https-v52lf?file=/App.tsx:1056-1103
使用 arrayfield, 当内容变更时，不会出发 onFieldValueChange，场景是其他字段依赖数组中的 min 和 max
codesandbox 的思路是希望可以内容变更的时候触发 onFieldValueChange 来更新依赖字段
![image](https://user-images.githubusercontent.com/24842116/127766916-c6ca6e69-7105-49e7-b5b7-a4a056bdd011.png)

还有一个思路：依赖字段设置 dependencies，同样无法触发(https://codesandbox.io/s/great-https-v52lf?file=/App.tsx:679-739)，还是需要指定字段比如 string_array.0.x，但是场景是需要从 string_array 中的 min 和 max 来判断，[这里](https://github.com/alibaba/formily/discussions/1620)有提到深度 clone，但是没有找到可以重制内容的 hook

**Comment by janryWang** [comment, marked as answer]
ArrayField操作是引用操作，不是赋值操作，所以是监听不到变化的，你可以用onFieldInputValueChange

**Reply by CandenWang** [reply]
为啥我用onFieldInputValueChange还是不能监听到变化，只有array数量改变时才能监听到变化，array里面string_array.0.x无法触发effect

**Reply by HuColin** [reply]
对于 ArrayField 中 length 的变化，如 insert remove 等操作，会触发 onInput 事件，进而触发 `LifeCycleTypes.ON_FIELD_INPUT_VALUE_CHANGE` `LifeCycleTypes.ON_FORM_INPUT_CHANGE` 对于数组的非直接对 length 的操作，还会触发 `LifeCycleTypes.ON_FORM_VALUES_CHANGE`
而对于 ArrayField 中每一项内容变化，同时还会触发 `LifeCycleTypes.ON_FIELD_VALUE_CHANGE`，此时触发的 field.path 是 string_array.0.x
以上结论在 `createForm effects` 里面得出来的

**Reply by Gllidan** [reply]
> 对于 ArrayField 中 length 的变化，如 insert remove 等操作，会触发 onInput 事件，进而触发 `LifeCycleTypes.ON_FIELD_INPUT_VALUE_CHANGE` `LifeCycleTypes.ON_FORM_INPUT_CHANGE` 对于数组的非直接对 length 的操作，还会触发 `LifeCycleTypes.ON_FORM_VALUES_CHANGE` 而对于 ArrayField 中每一项内容变化，同时还会触发 `LifeCycleTypes.ON_FIELD_VALUE_CHANGE`，此时触发的 field.path 是 string_array.0.x 以上结论在 `createForm effects` 里面得出来的

目前刚好在定位这里的问题，如上面说的，比如对数组做splice的删除操作，那会引起索引数值更新，会触发form的响应式更新，然后走initialValuesChange、formValuesChange逻辑。感觉这里是可以做层优化？因为实际会根据数组长度，重复触发valuesChange订阅执行
